cluster-autoscaler/debuggingsnapshot/debugging_snapshotter.go

```diff
@@ -0,0 +1,225 @@
+
+// IsDataCollectionAllowed encapsulate the check to know if data collection is currently active
+// The should be used by setters and by any function that is contingent on data collection State
+// before doing extra processing.
+// e.g. If you want to pre-process a particular State in cloud-provider for snapshot
+// you should check this func in the loop before doing that extra processing
+func (d *DebuggingSnapshotterImpl) IsDataCollectionAllowed() bool {
+	statesAllowed := map[DebuggingSnapshotterState]bool{
+		DATA_COLLECTED:        true,
+		START_DATA_COLLECTION: true,
+	}
+
+	return statesAllowed[*d.State]
+}
+
+// StartDataCollection changes the State when the trigger has been enabled
+// to start data collection. To be done at the start of the runLoop to allow for consistency
+// as the trigger can be called mid-loop leading to partial data collection
+func (d *DebuggingSnapshotterImpl) StartDataCollection() {
+	d.Mutex.Lock()
+	defer d.Mutex.Unlock()
+	if *d.State == TRIGGER_ENABLED {
+		*d.State = START_DATA_COLLECTION
+		klog.Infof("Trigger Enabled for Debugging Snapshot, starting data collection")
+		d.DebuggingSnapshot.SetStartTimestamp(time.Now().In(time.UTC))
+	}
+}
+
+// Flush is the impl for DebuggingSnapshotter.Flush
+// It checks if any data has been collected or data collection failed
+func (d *DebuggingSnapshotterImpl) Flush() {
+	d.Mutex.Lock()
+	defer d.Mutex.Unlock()
+
+	// Case where Data Collection was started but no data was collected, needs to
+	// be stated as an error and reset to pre-trigger State
+	if *d.State == START_DATA_COLLECTION {
+		klog.Errorf("No data was collected for the snapshot in this loop. So no snapshot can be generated.")
+		d.DebuggingSnapshot.SetErrorMessage("Unable to collect any data")
+		d.Trigger <- true
+		return
+	}
+
+	if *d.State == DATA_COLLECTED {
+		d.Trigger <- true
+	}
+}
+
+// SetNodeGroupInfo is the setter for Node Group Info
+// All filtering/prettifying of data should be done here.
+func (d *DebuggingSnapshotterImpl) SetNodeGroupInfo(nodeInfos []*framework.NodeInfo) {
+	d.Mutex.Lock()
+	defer d.Mutex.Unlock()
+	if !d.IsDataCollectionAllowed() {
+		return
+	}
+	klog.Infof("NodeGroupInfo is being set for the debugging snapshot")
+	d.DebuggingSnapshot.SetNodeGroupInfo(nodeInfos)
+	*d.State = DATA_COLLECTED
+}
+
+// SetUnscheduledPodsCanBeScheduled is the setter for UnscheduledPodsCanBeScheduled
+func (d *DebuggingSnapshotterImpl) SetUnscheduledPodsCanBeScheduled(podList []*v1.Pod) {
+	d.Mutex.Lock()
+	defer d.Mutex.Unlock()
+	if !d.IsDataCollectionAllowed() {
+		return
+	}
+	klog.Infof("UnscheduledPodsCanBeScheduled is being set for the debugging snapshot")
+	d.DebuggingSnapshot.SetUnscheduledPodsCanBeScheduled(podList)
+	*d.State = DATA_COLLECTED
+}
+
+// Cleanup clears the internal data sets of the cluster
+func (d *DebuggingSnapshotterImpl) Cleanup() {
+	// The new data bean is set by value, to avoid changing the reference
+	d.DebuggingSnapshot.Cleanup()
+	d.Trigger <- false
```

Comment:
This is going to hang until there's an incoming request. I'd suggest using `r.Context().WithCancel()` in the request handler, `select{}` on both trigger and `ctx.Done()`, and just calling the cancel function here.

Reply:
done